### ISSUE TITLE
Check for open ports

### ISSUE THREAD

**Issue by mbitokhov** [woodruffw/twa#18]
This might be outside the scope of twa, but it'd be informative if twa provided information about open, filtered, or closed ports on the server

Here's a list of potential ports
- 3306 (mysql)
- 9000 (default php-fpm port)
- 11211 (memcached)
- 8080 (prod shouldn't have dev environments)

That's as far as I can think of right now. But there's certainly more

**Comment by woodruffw** [comment]
Great idea! @ElwoodBlues99 began some initial work on this in #14, although I'd like to break it own into its own PR to keep things simple.

Ref: https://github.com/woodruffw/twa/pull/14#discussion_r217934455

**Comment by DLewis01** [comment]
8080 is also tomcat. You'd have to hit the open port, send some data and check what is returned to see what is actually running on that port.  You could have a set of default (usual) ports. 
I like the idea of detecting memecached and other proxies/CDNs which cause no end of trouble.

**Comment by woodruffw** [comment]
In the interest of simplicity, detecting exactly *what* is running on common development/backend ports might be out of scope. Instead, it'll probably be sufficient to `FAIL` (or at least `WARN`) if *anything at all* is listening for HTTP on them.

**Comment by DLewis01** [comment]
I'm in favour of putting in nmap. If you are examining a web server, I want one tool I can run that will will give me everything I need to know about it. Ports, misconfigurations, data leaks via headers, etc.

**Comment by woodruffw** [comment]
>  I want one tool I can run that will will give me everything I need to know about it.

I agree that a tool like this would be nice to have, but I'm hesitant to make it this one (as I said in another issue, I'd like to stay true to the "t" in `twa`). For the time being, I'd be more than happy to accept PRs that check for common ports used for development/backend HTTP services, like `8080`, `4567`, and `8443`.

I'll also leave this issue open so that others can find it and comment on adding more involved checks (for memcached, mySQL, etc). If we can get a broad consensus on those being desirable (and people bring forward implementations that don't substantially increase the dependencies/complicate the script), then I'll be happy to review those and bring them in as well.

**Comment by mbitokhov** [comment]
If I find some time over the weekend, I'll see if I can write something up. Be warned though, my bash-fu isn't that good

It doesn't seem too difficult to set up though. twa already has nc as a dependency, so all we have to do is check the output of `nc -z "${domain} 8443"`. Feature detection might be a bit more difficult, but doable. _If_ twa were to support mysql, we can check the output of
```bash
$ cat < "/dev/tcp/${domain}/3306"
> b
> 5.5.5-10.1.26-MariaDB-0+deb
```

**Comment by woodruffw** [comment]
Here's a (partial) list of well-known backend/development HTTP ports:

* 4443
* 4567
* 8000
* 8080
* 8081
* 8443

**Comment by mbitokhov** [comment]
Just for the sake of completeness, there's also 3000, which is nodejs and ruby on rails
